### ISSUE TITLE
Transactions

### ISSUE THREAD

**Issue by bryanburgers** [coderhaoxin/pg-then#17]
I need transactions in my current project. Is this something you could see added to pg-then, or should it be a new module?

I'm thinking the interface could be something like

```js
const pg = require('pg-then')

const pool = pg.Pool('postgres://username:password@localhost/database')

pool.transactional(tx => {
  return tx.query('UPDATE account SET amount = amount + 10 WHERE id = 1')
    .then(res => {
      return tx.query('UPDATE account SET amount = amount - 10 WHERE id = 2')
    });
});

// or concurrently

pool.transactional(tx => {
  const q1 = tx.query('UPDATE account SET amount = amount + 10 WHERE id = 1');
  const q2 = tx.query('UPDATE account SET amount = amount - 10 WHERE id = 2');
  return Promise.all([q1, q2]);
});
```

The function passed to `transactional` would return a promise, and if that promise resolved, the transaction would be committed; rejected the transaction would be rolled back. Rollbacks could also be explicit:

```js
pool.transactional(tx => {
  return tx.query('UPDATE account SET amount = amount + 100000')
    .then(res => {
      return tx.rollback();
    });
});
```

Let me know if this is something I should submit a pull request for, or should put in a different module.

**Comment by coderhaoxin** [comment]
SGTM, PR welcome!

**Comment by vitaly-t** [comment]
This will not work. It cost me a month of efforts when I was working on [pg-promise](https://github.com/vitaly-t/pg-promise) to understand the nature of the problem properly and to create a solution...

The problem:

You cannot use `Promise.all` in the context of an open database connection, because just as the method rejects, the connection closes, while the rest of queries will continue executing against a closed connection, which in turn creates all kinds of asynchronous problems.

The solution:

You need to rely on a promise method that settles all the queries, before trying to release the connection. This is why I ended up writing and using method [batch](https://github.com/vitaly-t/spex/blob/master/docs/code/batch.md) which can do all that well. And this isn't in the standard Promise protocol.

**Comment by bryanburgers** [comment]
That's a good point. I'm interested in learning more.

Is it just the reject immediately semantics of `Promise.all()` that cause issues? Would [Q's allSettled](https://github.com/kriskowal/q/wiki/API-Reference#promiseallsettled) method get around this?

Or does it make sense to wait to finish the transaction until both (1) the userland work function to `transactional` has finished, _and_ (2) any query started within the transaction has finished? Or will that also have subtle issues that I'm not thinking of?

Thanks for the heads up!

**Comment by vitaly-t** [comment]
>Is it just the reject immediately semantics of Promise.all() that cause issues? 
Yes. Once it rejects, the connection closes, but the rest of queries will try to execute.

>Would Q's allSettled method get around this?
Yes.

>Or does it make sense to wait to finish the transaction until both (1) the userland work function to transactional has finished, and (2) any query started within the transaction has finished? Or will that also have subtle issues that I'm not thinking of?

That's what `allSettled` would be doing on the array of queries. You must not return the transaction result till `allSettled` has returned.

And your callback returns first, only then you settle the queries, that's how async IO works.

Subtle issues? No.

**Comment by bryanburgers** [comment]
Did you find a good way to test this to make sure you were doing things right? Would these get what you need?

```js
function delay(timout) {
  return new Promise(resolve => {
    setTimeout(resolve, timeout);
  });
}

// Test a query that gets started before and finishes after an error doesn't wreak havoc
pool.transactional(tx => {
  const q1 = tx.query('SELECT pg_sleep(20)');
  const q2 = delay(1000).then(() => throw new Error("ERROR"));
  return Promise.all([q1, q2]);
});

// Test that trying to run a query AFTER another query has already rejected won't wreak havoc
pool.transactional(tx => {
  const q1 = delay(1000).then(() => tx.query('SELECT 1');
  const q2 = Promise.reject(new Error("ERROR"));
  return Promise.all([q1, q2]);
});

// Test that an error reported from the server on the connection doesn't sink us
pool.transactional(tx => {
  const q1 = tx.query('SELECT pg_sleep(20)');
  const q2 = delay(1000).then(() => tx.query('SELECT * FROM nonexistanttable'));
  return Promise.all([q1, q2]);
});
```

Are there any other tests that I need to make sure work?

If I assume `Promise.all` is a good idea, I could see this jumping up to bite other people too, so I would like to get `pool.transactional` to handle as much as possible.

I own the `tx` parameter, so in the above case #2, I can make can make sure that a query called after I've already rolled back just doesn't get run. In case #1, I can also just keep track of any queries that haven't completed yet when the work promise is rejected, and wait for them.

Am I on the right track, or is this always going to be futile?

**Comment by vitaly-t** [comment]
Doesn't look right. The kind of test you need:

A transaction that executes 2 queries via `Promise.all`. The first query must reject at once, and the second one execute with a delay. This will replicate the problem. I don't see such test on the list there.
This would be your negative test ;)

In [pg-promise](https://github.com/vitaly-t/pg-promise) I did very comprehensive testing. And on top of that, I added handling this situation, in case somebody tried to use `promise.all` instead of `batch` method, to throw for such failed queries: `Loose request outside an expired connection.`.

**Comment by vitaly-t** [comment]
Just couple of tests I did specifically for loose connections:

* https://github.com/vitaly-t/pg-promise/blob/9c1946cdff2d87ee153eba97e641468e8985871d/test/eventSpec.js#L341
* https://github.com/vitaly-t/pg-promise/blob/9c1946cdff2d87ee153eba97e641468e8985871d/test/eventSpec.js#L376

They would of course differ in your case, as the protocol differs.

**Comment by bryanburgers** [comment]
Thanks! I'll dig in when I have a chance.

**Comment by vitaly-t** [comment]
You are welcome!

It is fun watching somebody trying to rewrite what I did, repeating all the mistakes I did :)

You want more to the table? People do use nested transactions quite a bit. Within PostgreSQL those translate into save points, which in turn require automatic name generation and scope management.

Just in case you think you got to the difficult part, you aren't there yet ;)

:) :) :) :) :)
